test/tests/router_test.js

```diff
@@ -3327,6 +3327,660 @@ test("intermediateTransitionTo() forces an immediate intermediate transition tha
+      bazModelCount++;
+    }
+  };
+
+  handlers = {
+    foo: fooHandler,
+    bar: barHandler,
+    baz: bazHandler
+  };
+
+  transitionTo(router, '/foo');
+
+  assert.equal(url, '/baz');
+  assert.equal(fooModelCount, 1);
+  assert.equal(barModelCount, 1);
+  assert.equal(bazModelCount, 1);
+});
+
+test("Calling transitionTo after initial transition in validation hook should use updateUrl", function(assert) {
+  assert.expect(8);
+
+  map(assert, function(match) {
+    match("/foo").to('foo');
+    match("/bar").to('bar');
+  });
+
+  var fooModelCount = 0, barModelCount = 0;
+
+  router.updateURL = function(updateUrl) {
+    url = updateUrl;
+    assert.ok(true, "updateURL should be used");
+  };
+
+  router.replaceURL = function(replaceURL) {
+    url = replaceURL;
+    assert.ok(false, "replaceURL should not be used");
+  };
+
+  var fooHandler = {
+    model: function() {
+      fooModelCount++;
+      router.transitionTo('/bar');
+    }
+  };
+
+  var barHandler = {
+    model: function() {
+      barModelCount++;
+    }
+  };
+
+  handlers = {
+    foo: fooHandler,
+    bar: barHandler
+  };
+
+  transitionTo(router, '/bar');
+
+  assert.equal(url, '/bar');
+  assert.equal(barModelCount, 1, 'Bar model should be called once');
+  assert.equal(fooModelCount, 0, 'Foo model should not be called');
+
+  transitionTo(router, '/foo');
+
+  assert.equal(url, '/bar');
+  assert.equal(barModelCount, 2, 'Bar model should be called twice');
+  assert.equal(fooModelCount, 1, 'Foo model should be called once');
+});
+
+test("Calling transitionTo after initial transition in validation hook with multiple redirects should use updateUrl", function(assert) {
+  assert.expect(10);
+
+  map(assert, function(match) {
+    match('/foo').to('foo');
+    match('/bar').to('bar');
+    match('/baz').to('baz');
+  });
+
+  var fooModelCount = 0, barModelCount = 0, bazModelCount = 0;
+
+  router.updateURL = function(updateUrl) {
+    url = updateUrl;
+    assert.ok(true, "updateURL should be used");
+  };
+
+  router.replaceURL = function(replaceURL) {
+    url = replaceURL;
+    assert.ok(false, "replaceURL should not be used");
+  };
+
+  var fooHandler = {
+    model: function() {
+      fooModelCount++;
+      router.transitionTo('/bar');
+    }
+  };
+
+  var barHandler = {
+    model: function() {
+      barModelCount++;
+      router.transitionTo('/baz');
+    }
+  };
+
+  var bazHandler = {
+    model: function() {
+      bazModelCount++;
+    }
+  };
+
+  handlers = {
+    foo: fooHandler,
+    bar: barHandler,
+    baz: bazHandler
+  };
+
+  transitionTo(router, '/baz');
+
+  assert.equal(url, '/baz');
+  assert.equal(bazModelCount, 1, 'Baz model should be called once');
+  assert.equal(fooModelCount, 0, 'Foo model should not be called');
+  assert.equal(barModelCount, 0, 'Bar model should not be called');
+
+  transitionTo(router, '/foo');
+
+  assert.equal(url, '/baz');
+  assert.equal(bazModelCount, 2, 'Baz model should be called thrice');
```

Comment:
Something is wrong on this line, unsure whether it's the message or the expectation.